File: leetCode/easy/2194-cellsInARangeOnAnExcelSheet.jl
function cellsInARangeOnAnExcelSheet(s::String)
  alphabet = collect('A':'Z')

  alphabetIdxMap = Dict{String,Int64}()
  idxAlphabetMap = Dict{Int64,String}()

  for (idx, char) in enumerate(alphabet)
    str = string(char)
    alphabetIdxMap[str] = idx
    idxAlphabetMap[idx] = str
  end

  startCell, endCell = split(s, ":")
  startCell = string(startCell)
  endCell = string(endCell)

  startingCol, startingRow = split(startCell, "")
  startingCol = string(startingCol)
  startingRow = string(startingRow)

  endingCol, endingRow = split(endCell, "")
  endingCol = string(endingCol)
  endingRow = string(endingRow)

  colHeight = parse(Int, endingRow) - parse(Int, startingRow)
  startColNum = get(alphabetIdxMap, startingCol, "")
  endColNum = get(alphabetIdxMap, endingCol, "")
  rowDepth = endColNum - startColNum

  colHeightCounter = 0
  rowDepthCounter = 0
  colCharIdxStart = startColNum
  rowNumStart = parse(Int, startingRow)
  cellsResult::Array{String,1} = []

  while rowDepthCounter <= rowDepth
    while colHeightCounter <= colHeight
      cellStr = ""
      colChar = get(idxAlphabetMap, colCharIdxStart, "")
      cellStr *= colChar
      cellStr *= string(rowNumStart)
      push!(cellsResult, cellStr)
      rowNumStart += 1
      colHeightCounter += 1
    end

    colCharIdxStart += 1
    rowDepthCounter += 1
    colHeightCounter = 0
    rowNumStart = parse(Int, startingRow)
  end

  return cellsResult
end

export cellsInARangeOnAnExcelSheet

using Test

Test.@testset "cellsInARangeOnAnExcelSheet" begin
  Test.@test cellsInARangeOnAnExcelSheet("A1:B2") == ["A1", "A2", "B1", "B2"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:A1") == ["A1"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:B3") == ["A1", "A2", "A3", "B1", "B2", "B3"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:C3") == ["A1", "A2", "A3", "B1", "B2", "B3", "C1", "C2", "C3"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:A3") == ["A1", "A2", "A3"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:C1") == ["A1", "B1", "C1"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:C2") == ["A1", "A2", "B1", "B2", "C1", "C2"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:B4") == ["A1", "A2", "A3", "A4", "B1", "B2", "B3", "B4"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:C4") == ["A1", "A2", "A3", "A4", "B1", "B2", "B3", "B4", "C1", "C2", "C3", "C4"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:A4") == ["A1", "A2", "A3", "A4"]
  Test.@test cellsInARangeOnAnExcelSheet("A1:C3") == ["A1", "A2", "A3", "B1", "B2", "B3", "C1", "C2", "C3"]
end

File: leetCode/easy/2194-cellsInARangeOnAnExcelSheet.ts
import { assertEquals } from "https://deno.land/std@0.182.0/testing/asserts.ts";
import { describe, it } from "https://deno.land/std@0.182.0/testing/bdd.ts";

function cellsInARangeOnAnExcelSheet(s: string) {
  const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  const [alphabetIdxMap, idxAlphabetMap] = alphabet.split("").reduce(
    (maps: [Map<string, number>, Map<number, string>], char, idx) => {
      maps[0].set(char, idx + 1);
      maps[1].set(idx + 1, char);

      return maps;
    },
    [new Map(), new Map()]
  );

  const [startCell, endCell] = s.split(":");
  const [startingCol, startingRow] = startCell.split("");
  const [endingCol, endingRow] = endCell.split("");

  const colHeight = Number(endingRow) - Number(startingRow);
  const startColNum = alphabetIdxMap?.get(startingCol) ?? 0;
  const endColNum = alphabetIdxMap.get(endingCol) ?? 0;
  const rowDepth = endColNum - startColNum;

  let colHeightCounter = 0;
  let rowDepthCounter = 0;
  let colCharIdxStart = startColNum;
  let rowNumStart = Number(startingRow);
  const cellsResult: string[] = [];

  while (rowDepthCounter <= rowDepth) {
    while (colHeightCounter <= colHeight) {
      let cellStr = "";
      const colChar = idxAlphabetMap.get(colCharIdxStart) ?? "";
      cellStr += colChar;
      cellStr += rowNumStart.toString();
      cellsResult.push(cellStr);
      rowNumStart += 1;
      colHeightCounter += 1;
    }

    colCharIdxStart += 1;
    rowDepthCounter += 1;
    colHeightCounter = 0;
    rowNumStart = Number(startingRow);
  }

  return cellsResult;
}

export { cellsInARangeOnAnExcelSheet };

describe("cellsInARangeOnAnExcelSheet", () => {
  it("1. should return the cells in a range on an excel sheet", () => {
    const s = "A1:B2";
    assertEquals(cellsInARangeOnAnExcelSheet(s), ["A1", "A2", "B1", "B2"]);
  });

  it("2. should return the cells in a range on an excel sheet", () => {
    const s = "A1:F1";
    assertEquals(cellsInARangeOnAnExcelSheet(s), [
      "A1",
      "B1",
      "C1",
      "D1",
      "E1",
      "F1",
    ]);
  });
});
